### PR TITLE
feat(PLT-271): add cache to finance core api

### PR DESCRIPTION
**Pull request by barkerxuan**
This PR adds cache to finance core API

### CODE REVIEW

File: src/FinanceCoreApi.php [hunk 1]
```diff
@@ -8,29 +8,43 @@
...
 
+    /** @var CacheItemPoolInterface|null */
+    protected $cacheItemPool;
+
+    /**
+     * @param \Psr\Log\LoggerInterface $logger
+     * @param \BrighteCapital\Api\BrighteApi $brighteApi
+     * @param CacheItemPoolInterface $cache
+     */
+    public function __construct(LoggerInterface $logger, BrighteApi $brighteApi, CacheItemPoolInterface $cache)
```
Comment (darkbluesun) [inline src/FinanceCoreApi.php:28]: in order to not break clients, you should make CacheItemPoolInterface nullable with a `?`

File: src/FinanceCoreApi.php [hunk 2]
```diff
@@ -181,4 +192,26 @@ private function checkIfContainsError(string $function, ResponseInterface $respo
         }
         return $body;
     }
+
+    private function getCached(string $functionName, array $parameters, array $body)
+    {
+        $key = implode('_', [$functionName, implode('_', $parameters)]);
+        if ($cachedItem = $this->cacheItemPool->getItem($key)) {
```
Comment (darkbluesun) [inline src/FinanceCoreApi.php:199]: you should handle if `$this->cacheItemPool` is null

**Review by darkbluesun — CHANGES_REQUESTED** [review]
Hi Barker
This is good work, but I have two critical things that need fixing first, inline comments below
Also, had you considered adding this functionality to the BrighteApi class? and if so, why did you decide against it?

**Comment by barkerxuan** [comment]
@darkbluesun I think it is doable, but I was under the impression that BrighteApi serves as a shared class that will be used by other classes. The cached function is quite specific to the finance core(Graphql) as we cache the response of post requests. We could add a new function like `cachedPost` in BrighteApi, do you think that is a better approach?

**Comment by darkbluesun** [comment]
Yeah if you don't mind barker, I think a cachedPost in brighteApi would be much better because I think we will get more graphql apis over time and this will become a thing. The only way it's specific to the financecore api is just that it's post instead of get.

File: src/BrighteApi.php [hunk 3]
```diff
@@ -193,6 +195,56 @@ public function post(
         return $this->doRequest('POST', $path, $query, $body, $headers, $auth);
     }
 
+    public function cachedPost(
+        string $functionName,
```
Comment (darkbluesun) [inline src/BrighteApi.php:199]: maybe instead of using these as input, just take the string $key as input here, as with getCached.

**Review by darkbluesun — APPROVED** [review]
Looking good 👍

File: src/BrighteApi.php [hunk 4]
```diff
@@ -193,6 +195,56 @@ public function post(
...
+        $response = $this->doRequest('POST', $path, $query, $body, $headers, $auth);
+
+        $responseBody = $this->checkIfContainsError($functionName, $response);
+        if ($responseBody === null) {
+            return null;
+        }
+
+        if ($this->cacheItemPool) {
+            $item = new CacheItem($key, true, $responseBody);
+            $expires = new \DateInterval('PT' . strtoupper("15m"));
```
Comment (hbhrugubanda) [inline src/BrighteApi.php:221]: question: out of curiosity, how did we come up with 15 mins?

File: tests/FinanceCoreApiTest.php [hunk 5]
```diff
@@ -206,105 +205,31 @@ public function testgetFinancialProduct(): void
...
-    * @covers ::checkIfContainsError
-    * @covers ::logGraphqlResponse
-    */
-    public function testgetFinancialProductConfigFailWhenServerError(): void
+     * @covers ::__construct
+     * @covers ::getFinancialProductConfig
+     * @covers ::getFinancialProductConfigFromResponse
+     * @covers ::createGetFinancialProductConfigQuery
+     */
+    public function testgetFinancialProductConfigWhenReturnsNull(): void
```
Comment (hbhrugubanda) [inline tests/FinanceCoreApiTest.php:213]:
```suggestion
    public function testGetFinancialProductConfigWhenReturnsNull(): void
```

File: tests/FinanceCoreApiTest.php [hunk 6]
```diff
@@ -206,105 +205,31 @@ public function testgetFinancialProduct(): void
...
-    * @covers ::getFinancialProduct
-    * @covers ::checkIfContainsError
-    * @covers ::logGraphqlResponse
-    */
-    public function testgetFinancialProductFailWhenServerError(): void
+     * @covers ::__construct
+     * @covers ::getFinancialProduct
+     * @covers ::getFinancialProductConfigFromResponse
+     */
+    public function testgetFinancialProductWhenReturnsNull(): void
```
Comment (hbhrugubanda) [inline tests/FinanceCoreApiTest.php:227]:
```suggestion
    public function testGetFinancialProductWhenReturnsNull(): void
```

**Review by hbhrugubanda — APPROVED** [review]
Looks good! 💯  No blocking comments, however please wait for Chris to approve as well ✅

**Review by darkbluesun — APPROVED** [review]

**Pull request merged**